smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java

```diff
@@ -0,0 +1,349 @@
+     */
+    public Optional<ArtifactDefinitions> getArtifactDefinitions() {
+        return Optional.ofNullable(artifactDefinitions);
+    }
+
+    /**
+     * Gets this TraceFile's Shapes map.
+     * The shapes {@link Map} provides a mapping of absolute Smithy shape IDs to a list
+     * of shape link objects. A single Smithy shape can be responsible for generating
+     * multiple components in the target artifact.
+     *
+     * @return a Map from ShapeIDs to a list of ShapeLink's that represents the contents of the
+     * shapes tag in the trace file.
+     */
+    public Map<ShapeId, List<ShapeLink>> getShapes() {
+        return shapes;
+    }
+
+    /**
+     * Take this object and create a builder that contains all of the
+     * current property values of this object.
+     *
+     * @return a builder for type T
+     */
+    @Override
+    public Builder toBuilder() {
+        return builder()
+                .artifact(artifactMetadata)
+                .smithyTrace(smithyTrace)
+                .definitions(artifactDefinitions)
+                .shapes(shapes);
+    }
+
+    /**
+     * Builder for constructing TraceFile's from scratch.
+     */
+    public static final class Builder implements SmithyBuilder<TraceFile> {
+
+        private final Map<ShapeId, List<ShapeLink>> shapes = new HashMap<>();
+        private String smithyTrace = SMITHY_TRACE_VERSION;
+        private ArtifactDefinitions artifactDefinitions;
+        private ArtifactMetadata artifactMetadata;
+
+        /**
+         * @return The TraceFile.
+         */
+        public TraceFile build() {
+            return new TraceFile(this);
+        }
+
+        /**
+         * @param smithyTrace Trace file version number.
+         * @return This builder.
+         */
+        public Builder smithyTrace(String smithyTrace) {
+            this.smithyTrace = smithyTrace;
+            return this;
+        }
+
+        /**
+         * @param artifactDefinitions Trace file definitions.
+         * @return This builder.
+         */
+        public Builder definitions(ArtifactDefinitions artifactDefinitions) {
+            this.artifactDefinitions = artifactDefinitions;
+            return this;
+        }
+
+        /**
+         * @param artifactMetadata Trace file ArtifactMetadata.
+         * @return This builder.
+         */
+        public Builder artifact(ArtifactMetadata artifactMetadata) {
+            this.artifactMetadata = artifactMetadata;
+            return this;
+        }
+
+        /**
+         * Adds a ShapeLink to this ShapeId in the TraceFile's shapes map.
+         *
+         * @param id   ShapeId
+         * @param link ShapeLink corresponding to ShapeId
+         * @return This builder.
+         */
+        public Builder addShapeLink(ShapeId id, ShapeLink link) {
+            if (!this.shapes.containsKey(id)) {
+                this.shapes.put(id, new ArrayList<>());
+            }
+            this.shapes.get(id).add(link);
```

Comment:
You can use `.getOrDefault(id, new ArrayList<>())` here to avoid doing the pre-check by yourself.